### ISSUE TITLE
Players

### ISSUE THREAD

**Issue by Proximyst** [Proximyst/USSRM#10]
The following points are listed on a player:

From entities:

  * Bans
  * History
  * Permissions

From packets:

  * Exposed network manager
  * Easy to use

From cache:

  * Player cache

**Comment by Dot-Rar** [comment]
Should we really be handling bans? I'd say let a plugin do that.

**Comment by yannicklamprecht** [comment]
I would recommend outsourcing the bans into a service, implemented by a plugin.

**Comment by Proximyst** [comment]
That's fair enough. I at least find we should have some form for absolute base interface, which is then usable from any plugins, such that we don't have to go through anything Vault-like.

**Comment by Proximyst** [comment]
In that case, how about on the event of them joining, we add some form for "reason" for not being allowed on, just like they do in Spigot? E.g. having the following values, though in enum naming conventions?

  * Acceptation
  * Generic denial
  * Not whitelisted
  * Banned
  * Generic kick
  * Propagation of some sort
  * (BungeeCord/LilyPad integration) Not through proxy
  * No permission/maintenance 
  * Server shutting down
  * Error while fetching data in (async?) prelogin event

**Comment by Proximyst** [comment]
Thing is, interfaces are possible to be extended and implemented, but you'll always get the top level type you wish to get. When you try to fetch using e.g. `Server.instance.getService<PermissionService>()`, you'll get a `PermissionService?` back. This can either be null, an implementation of the service or more subtypes until it inevitably ends up on an implementation it wishes. If you wish to do so, you could check whether you're using that exact plugin with either `Server.instance.getService<PermissionService>()#.provider` (where # is !! or ? or some other form for check) or simply checking with an `is` or `==` (e.g. on objects) check.

**Comment by yannicklamprecht** [comment]
Yeah. I know. I think pulling methods up in later API versions could be done easily.
A good definition of an service interface can avoid the problem that it lacks functionality.
